python/tvm/topi/x86/tensor_intrin.py

```diff
@@ -348,3 +348,227 @@ def _instr(index):
         binds={data: a_buffer, kernel: b_buffer},
         default_buffer_params=buffer_params,
     )
+
+
+def dot_32x128x32_u8s8s32_sapphirerapids(LDA):
+    """
+    Int8 dot product by every 16x64 elements using AMX-TMUL Sapphire Rapids instructions.
+    The tdpxxd instruction takes two tile of uint8 and int8 datatype -- data[16][64] and
+    kernel[1][16][16][4] -- and computes a dot product of data[16][16] in int32 datatype.
+
+    (Physically, to efficiently leveraging the tile register, we constructing a 2x2 tiles
+    matmul which performs 32x128x32 in total)
+
+    The pseudo code is as follows:
+        for(k=0; k<2; k++){
+            for(n=0; n<2; n++){
+                tileload64(tmm_b, B)
+                for(m=0; m<2; m++){
+                    if(n==0)
+                        tileload64(tmm_a, A)
```

Comment:
Have you considered tensorizing at finer granularity? Meaning, instead of tensoring the whole load + compute as one tensor intrin, tensorize each load and compute separately. That could make the hard coded outer-loop factors (2 in this intrin) tunable.

Reply:
@masahi ,

I have a question related to your proposed here:

* Is there a TIR _desc equivalent for TOPI's body, reset, update ?
     topi intrin decl:  https://github.com/apache/tvm/blob/main/python/tvm/topi/x86/tensor_intrin.py#L140
* How the ```accum``` steps (very inner) can be succeeded by ```update``` (outer) or preceded by ```reset``` ?

I am looking for any TVM reference (code, example) for this, trying even finer granularity control of tensorization.

Reply:
If you are asking about tensorization in TIR, there is no need to have separate init, update etc. See https://github.com/apache/tvm/blob/main/python/tvm/tir/tensor_intrin/x86.py. The accum buffer is implicitly zero-initialized and we can direct do update via `+=`.

Reply:
> If you are asking about tensorization in TIR, there is no need to have separate init, update etc. See https://github.com/apache/tvm/blob/main/python/tvm/tir/tensor_intrin/x86.py. The accum buffer is implicitly zero-initialized and we can direct do update via `+=`.

@masahi ,
Thanks a lot ! I've noticed now "+="  for ```update``` purpose in TIR description.
Will also look @ MultiLevelTiling() (cuda only) idea for custom ISA having special ```init```, ```load_*```, ```store```.

Reply:
> Have you considered tensorizing at finer granularity? Meaning, instead of tensoring the whole load + compute as one tensor intrin, tensorize each load and compute separately. That could make the hard coded outer-loop factors (2 in this intrin) tunable.

Yes we considered to make it works like tensor core which mentioned here: https://github.com/apache/tvm/issues/4052, but the problem is AMX intrinsics in LLVM need user to assigned register(tmm) number by hand craft. If we seperate the load and compute, we have to leave the register assginment to schedule method, that would bring much redundancy work. So finnally we decide to follow BRGEMM and leave the whole smallest comute unit for tensorizing.

Reply:
@masahi ,
@Qianshui-Jiang ,

Don't want hold up the PR, I think it is fine to merge with this start AMX implementation.
I  leave here a possible way to workaround the rigidity of LLVM's AMX related intrinsic declaration:

* Use ```tvm.tir.call_extern``` insead of ```tvm.tir.call_llvm_intrin``` like [here](https://github.com/apache/tvm/blob/8551a5c71faaa2eedc8b1c49a39be0fe4688c021/python/tvm/topi/arm_cpu/mprofile/dsp/micro_kernel/gemm.py#L107-L109) .
* You can than pass argument (e.g. register number, etc ...) into the custom called function.
* You can use the argument inside that function and emit the AMX ISA as bytes: [ __asm(.word)](https://github.com/cbalint13/OLIMP/blob/main/demo/icebreaker/src/rv32-custom.h#L52-L53)

This is how it is done to emit custom part of RISC-V ISA, that is also too rigid using just the LLVM declarations.

Reply:
@cbalint13, Thanks a lot for your advices. 
Regarding that in TVM we already have DNNL codegen and some external call to DNNL, which is an off-the -shelf compute lib leverages SIMD ISA transparently.
The main focus in this PR is to add more native capability for TVM to generate AMX instructions. 
AFAIK, the LLVM features for AMX are also moving forward such as register allocation for intrinsics and auto vectorization.
Hence we choose to call LLVM intrinsics directly and hope to have more flexibility in the future.